### PR TITLE
Add support for ExcludedActions and ExcludedPrinicpals

### PR DESCRIPTION
**Pull request by Xynoclafe**
Author:    Gokul Premraj <gopremra@microsoft.com>

## Description

Added support for ExcludedActions and ExcludedPrincipals in BlueprintAssignment cmdlets

## Checklist

- [ ] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

File: src/Blueprint/Blueprint/Blueprint.format.ps1xml [hunk 1]
```diff
@@ -183,6 +183,22 @@
                 <Label>LockMode</Label>
                   <ScriptBlock>$_.Locks.Mode.ToString()</ScriptBlock>
               </ListItem>
+				<ListItem>
```
Comment (lwang2016) [inline src/Blueprint/Blueprint/Blueprint.format.ps1xml:186]: Can you fix indention?

File: src/Blueprint/Blueprint/Blueprint.format.ps1xml [hunk 2]
```diff
@@ -183,6 +183,22 @@
...
+					<Label>ExcludedActions</Label>
+					<ScriptBlock>
+						If (-not ([string]::IsNullOrEmpty($_.Locks.ExcludedActions))){
+						($_.Locks.ExcludedActions) -join "; "
```
Comment (lwang2016) [inline src/Blueprint/Blueprint/Blueprint.format.ps1xml:190]: @filiz, what delimiter do we usually use for joining list items?

File: src/Blueprint/Blueprint/Models/PSBlueprintAssignment.cs [hunk 3]
```diff
@@ -59,7 +59,12 @@ internal static PSBlueprintAssignment FromAssignment(Assignment assignment)
                 BlueprintId = assignment.BlueprintId,
                 ProvisioningState = PSAssignmentProvisioningState.Unknown,
                 Status = new PSAssignmentStatus(),
-                Locks = new PSAssignmentLockSettings {Mode = PSLockMode.None},
+                Locks = new PSAssignmentLockSettings
+                {
+                    Mode = PSLockMode.None,
+                    ExcludedActions = (List<string>)assignment.Locks.ExcludedActions,
```
Comment (lwang2016) [inline src/Blueprint/Blueprint/Models/PSBlueprintAssignment.cs:65]: nit, use ToList().

File: src/Blueprint/Blueprint/Blueprint.format.ps1xml (hunk 2 above)
Comment (filizt) [reply src/Blueprint/Blueprint/Blueprint.format.ps1xml:190]: It depends on the context. For this one, we considered comma, semi colon and line item and ended up using semi-colon. It'd be good to have your thoughts on that. @Xynoclafe could you send an email to Lei, Jorge and me showing how those 3 options would look in the output? We can update it if we thought some other delimiter looks better.

---
In reply to: [413184933](https://github.com/Azure/azure-powershell/pull/11661#discussion_r413184933) [](ancestors = 413184933)

File: src/Blueprint/Blueprint/Blueprint.format.ps1xml (hunk 1 above)
Comment (filizt) [reply src/Blueprint/Blueprint/Blueprint.format.ps1xml:186]: +1

---
In reply to: [413182628](https://github.com/Azure/azure-powershell/pull/11661#discussion_r413182628) [](ancestors = 413182628)

File: src/Blueprint/Blueprint/Models/PSBlueprintAssignment.cs (hunk 3 above)
Comment (Xynoclafe) [reply src/Blueprint/Blueprint/Models/PSBlueprintAssignment.cs:65]: Okay

File: src/Blueprint/Blueprint/Blueprint.format.ps1xml (hunk 1 above)
Comment (Xynoclafe) [reply src/Blueprint/Blueprint/Blueprint.format.ps1xml:186]: Not sure why it shows up that way. Looked fine on my IDE. Will check and fix it.

File: src/Blueprint/Blueprint/Blueprint.format.ps1xml (hunk 2 above)
Comment (Xynoclafe) [reply src/Blueprint/Blueprint/Blueprint.format.ps1xml:190]: Okay

File: src/Blueprint/Blueprint/Models/PSBlueprintAssignment.cs (hunk 3 above)
Comment (lwang2016) [reply src/Blueprint/Blueprint/Models/PSBlueprintAssignment.cs]: I thought we agreed to expose ExcludedPrincipals and ExcludedActions as array of strings and no type conversion is needed then.

File: src/Blueprint/Blueprint/Models/PSBlueprintAssignment.cs [hunk 4]
```diff
@@ -59,7 +60,12 @@ internal static PSBlueprintAssignment FromAssignment(Assignment assignment)
                 BlueprintId = assignment.BlueprintId,
                 ProvisioningState = PSAssignmentProvisioningState.Unknown,
                 Status = new PSAssignmentStatus(),
-                Locks = new PSAssignmentLockSettings {Mode = PSLockMode.None},
+                Locks = new PSAssignmentLockSettings
+                {
+                    Mode = PSLockMode.None,
+                    ExcludedActions = assignment.Locks.ExcludedActions.ToList(),
+                    ExcludedPrincipals = assignment.Locks.ExcludedPrincipals.ToList()
```
Comment (filizt) [inline src/Blueprint/Blueprint/Models/PSBlueprintAssignment.cs:67]: >assignment.Locks.ExcludedPrincipals.ToList() [](start = 41, length = 44)

What happens when ExcludedPrincipals is null?

File: src/Blueprint/Blueprint/Models/PSBlueprintAssignment.cs (hunk 3 above)
Comment (Xynoclafe) [reply src/Blueprint/Blueprint/Models/PSBlueprintAssignment.cs]: From what I saw, the SDK exposes ExcludedActions and ExcludedPrincipals as IList<string>. I wasn't sure if I could use an IList directly.

File: src/Blueprint/Blueprint/Models/PSBlueprintAssignment.cs (hunk 4 above)
Comment (Xynoclafe) [reply src/Blueprint/Blueprint/Models/PSBlueprintAssignment.cs:67]: Just realized that.

File: src/Blueprint/Blueprint/Models/PSBlueprintAssignment.cs (hunk 3 above)
Comment (lwang2016) [reply src/Blueprint/Blueprint/Models/PSBlueprintAssignment.cs]: Did you run into any errors if using IList directly?

Comment (Xynoclafe) [reply src/Blueprint/Blueprint/Models/PSBlueprintAssignment.cs]: I don't think I faced any errors. Just wasn't sure if it was a good practice.

File: src/Blueprint/Blueprint/Blueprint.format.ps1xml (hunk 1 above)
Comment (majastrz) [reply src/Blueprint/Blueprint/Blueprint.format.ps1xml]: It was tabs vs spaces btw.

---
In reply to: [413225736](https://github.com/Azure/azure-powershell/pull/11661#discussion_r413225736) [](ancestors = 413225736)

Comment (Xynoclafe) [reply src/Blueprint/Blueprint/Blueprint.format.ps1xml]: @majastrz Yep, took me a while to realize that.

**Pull request merged**